UCP: Use portable format string like PRIx64.

hiroyuki-sato opened the pull request:
## What

This PR uses a portable printf format like `PRIx64` instead of `%lx.`

## Why ?

Same as #5547, #5548 and #5549 reason. (for porting macOS)
For fixing https://github.com/hiroyuki-sato/ucx/issues/38

## How ?

Use `PTRx64` and `PTRxPTR` instead of `%lx`.

swx-jenkins3 commented:
Can one of the admins verify this patch?

hiroyuki-sato commented on src/ucp/proto/rndv.c:
```diff
@@ -1655,24 +1657,24 @@ static void ucp_rndv_dump(ucp_worker_h worker, uct_am_trace_type_t type,
-                 rndv_rtr_hdr->address);
+        snprintf(buffer, max, "RNDV_RTR sreq 0x%"PRIxPTR " rreq 0x%"PRIxPTR
+                 " address 0x%"PRIx64, rndv_rtr_hdr->sreq_ptr,
+                 rndv_rtr_hdr->rreq_ptr, rndv_rtr_hdr->address);
```
```C
    const ucp_rndv_rtr_hdr_t *rndv_rtr_hdr = data;

typedef struct {
    uintptr_t                 sreq_ptr; /* request on the rndv initiator side - sender */
    uintptr_t                 rreq_ptr; /* request on the rndv receiver side */
    uint64_t                  address;  /* holds the address of the data buffer on the receiver's side */
    size_t                    size;     /* size of the data to receive */
    size_t                    offset;   /* offset of the data in the recv buffer */
    /* packed rkeys follow */
} UCS_S_PACKED ucp_rndv_rtr_hdr_t;
```

hiroyuki-sato commented on src/ucp/rma/amo_send.c:
```diff
@@ -172,7 +172,8 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_atomic_op_nbx,
-        ucs_error("invalid atomic operation datatype: %zu", param->datatype);
+        ucs_error("invalid atomic operation datatype: %"PRIu64,
+                  param->datatype);
         return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM);
```
```C
UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_atomic_op_nbx,
//...
                 const ucp_request_param_t *param)



typedef struct {

    ucp_datatype_t datatype;

} ucp_request_param_t;


typedef uint64_t                         ucp_datatype_t;
```

hiroyuki-sato commented on src/ucp/rma/amo_send.c:
```diff
@@ -182,8 +183,8 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_atomic_op_nbx,
+                  " rkey %p to %s cb %p", opcode, buffer,
                   (param->op_attr_mask & UCP_OP_ATTR_FIELD_REPLY_BUFFER) ?
                   param->reply_buffer : NULL, param->datatype,
                   remote_addr, rkey, ucp_ep_peer_name(ep),
```
Same as above

yosefe commented:
ok to test

hiroyuki-sato commented on src/ucp/rma/amo_sw.c:
```diff
@@ -285,7 +285,8 @@ static void ucp_amo_sw_dump_packet(ucp_worker_h worker, uct_am_trace_type_t type
+                 "ATOMIC_REQ [addr 0x%"PRIx64" len %u reqptr "
+                 "0x%"PRIxPTR" ep 0x%"PRIxPTR" op %d]",
                  atomich->address, atomich->length, atomich->req.reqptr,
                  atomich->req.ep_ptr, atomich->opcode);
```
```C
    const ucp_atomic_req_hdr_t *atomich;

typedef struct {
    uint64_t                  address;
    ucp_request_hdr_t         req; /* NULL if no reply */
    uint8_t                   length;
    uint8_t                   opcode;
} UCS_S_PACKED ucp_atomic_req_hdr_t;

typedef struct {
    uintptr_t                 ep_ptr;
    uintptr_t                 reqptr;
} UCS_S_PACKED ucp_request_hdr_t;

```

hiroyuki-sato commented on src/ucp/rma/rma_sw.c:
```diff
@@ -263,23 +263,24 @@ static void ucp_rma_sw_dump_packet(ucp_worker_h worker, uct_am_trace_type_t type
-                 puth->ep_ptr);
+        snprintf(buffer, max, "PUT [addr 0x%"PRIx64" ep_ptr 0x%"PRIxPTR"]",
+                 puth->address, puth->ep_ptr);
         header_len = sizeof(*puth);
```
```C
    const ucp_put_hdr_t *puth;

typedef struct {
    uint64_t                  address;
    uintptr_t                 ep_ptr;
} UCS_S_PACKED ucp_put_hdr_t;
```

hiroyuki-sato commented on src/ucp/rma/rma_sw.c:
```diff
@@ -263,23 +263,24 @@ static void ucp_rma_sw_dump_packet(ucp_worker_h worker, uct_am_trace_type_t type
+        snprintf(buffer, max, "GET_REQ [addr 0x%"PRIx64" len %"PRIu64
+                 " reqptr 0x%"PRIxPTR" ep 0x%"PRIxPTR"]", geth->address,
+                 geth->length, geth->req.reqptr, geth->req.ep_ptr);
         return;
```
```C
    const ucp_get_req_hdr_t *geth;

typedef struct {
    uint64_t                  address;
    uint64_t                  length;
    ucp_request_hdr_t         req;
} UCS_S_PACKED ucp_get_req_hdr_t;
```

hiroyuki-sato commented on src/ucp/rma/rma_sw.c:
```diff
@@ -263,23 +263,24 @@ static void ucp_rma_sw_dump_packet(ucp_worker_h worker, uct_am_trace_type_t type
         reph = data;
-        snprintf(buffer, max, "GET_REP [reqptr 0x%lx]", reph->req);
+        snprintf(buffer, max, "GET_REP [reqptr 0x%"PRIxPTR"]", reph->req);
         header_len = sizeof(*reph);
```
```C
    const ucp_rma_rep_hdr_t *reph;

typedef struct {
    uintptr_t                 req;
} UCS_S_PACKED ucp_rma_rep_hdr_t;
```

yosefe commented on src/ucp/core/ucp_request.c:
```diff
@@ -268,16 +269,17 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_request_memory_reg,
     default:
         status = UCS_ERR_INVALID_PARAM;
-        ucs_error("Invalid data type %lx", datatype);
+        ucs_error("Invalid data type %"PRIx64, datatype);
```
pls add 0x

hiroyuki-sato replied:
done

hiroyuki-sato commented on src/ucp/rma/rma_sw.c:
```diff
@@ -263,23 +263,24 @@ static void ucp_rma_sw_dump_packet(ucp_worker_h worker, uct_am_trace_type_t type
         cmplh = data;
-        snprintf(buffer, max, "CMPL [ep_ptr 0x%lx]", cmplh->ep_ptr);
+        snprintf(buffer, max, "CMPL [ep_ptr 0x%"PRIxPTR"]", cmplh->ep_ptr);
         return;
```
```C
    const ucp_cmpl_hdr_t *cmplh;

typedef struct {
    uintptr_t                 ep_ptr;
} UCS_S_PACKED ucp_cmpl_hdr_t;
```

yosefe commented on src/ucp/core/ucp_request.inl:
```diff
@@ -558,7 +558,7 @@ ucp_request_recv_data_unpack(ucp_request_t *req, const void *data,
 
     default:
-        ucs_fatal("unexpected datatype=%lx", req->recv.datatype);
+        ucs_fatal("unexpected datatype=%"PRIx64, req->recv.datatype);
```
0x

hiroyuki-sato replied:
done

yosefe commented on src/ucp/dt/dt.inl:
```diff
@@ -107,7 +107,7 @@ ucp_dt_unpack_only(ucp_worker_h worker, void *buffer, size_t count,
 
     default:
-        ucs_fatal("unexpected datatype=%lx", datatype);
+        ucs_fatal("unexpected datatype=%"PRIx64, datatype);
```
0x

hiroyuki-sato replied:
done

yosefe commented on src/ucp/proto/rndv.c:
```diff
@@ -986,7 +987,8 @@ static void ucp_rndv_do_rkey_ptr(ucp_request_t *rndv_req, ucp_request_t *rreq,
          */
-        ucs_fatal("failed to find a lane to access remote memory domains 0x%lx",
+        ucs_fatal("failed to find a lane to access remote memory domains "
+                  "0x%"PRIx64,
```
unite rkey->md_map and PRIx64 to same line

hiroyuki-sato replied:
done

hiroyuki-sato commented on src/ucp/tag/eager_rcv.c:
```diff
@@ -474,12 +474,12 @@ static void ucp_eager_dump(ucp_worker_h worker, uct_am_trace_type_t type,
-                 ucs_status_string(rep_hdr->status));
+        snprintf(buffer, max, "EGRS_A request 0x%"PRIx64" status '%s'",
+                 rep_hdr->reqptr, ucs_status_string(rep_hdr->status));
         header_len = sizeof(*rep_hdr);
```
```C
    const ucp_reply_hdr_t *rep_hdr               = data;

typedef struct {
    uint64_t                  reqptr;
    ucs_status_t              status;
} UCS_S_PACKED ucp_reply_hdr_t;
```

yosefe commented on src/ucp/rma/amo_send.c:
```diff
@@ -172,7 +172,8 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_atomic_op_nbx,
         op_size = sizeof(uint32_t);
     } else {
-        ucs_error("invalid atomic operation datatype: %zu", param->datatype);
+        ucs_error("invalid atomic operation datatype: %"PRIu64,
```
0x%"PRIx64

hiroyuki-sato replied:
done

hiroyuki-sato commented on src/ucp/tag/eager_rcv.c:
```diff
@@ -474,12 +474,12 @@ static void ucp_eager_dump(ucp_worker_h worker, uct_am_trace_type_t type,
-        snprintf(buffer, max, "EGRS_A_O tag %"PRIx64" ep_ptr 0x%lx",
+        snprintf(buffer, max, "EGRS_A_O tag %"PRIx64" ep_ptr 0x%"PRIxPTR,
                  off_rep_hdr->sender_tag, off_rep_hdr->ep_ptr);
         header_len = sizeof(*rep_hdr);
```
```C
    const ucp_offload_ssend_hdr_t *off_rep_hdr   = data;

typedef struct {
    uintptr_t         ep_ptr;
    ucp_tag_t         sender_tag;
} UCS_S_PACKED ucp_offload_ssend_hdr_t;
```

yosefe commented on src/ucp/rma/amo_send.c:
```diff
@@ -182,8 +183,8 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_atomic_op_nbx,
     ucs_trace_req("atomic_op_nbx opcode %d buffer %p result %p "
-                  "datatype %zu remote_addr %"PRIx64" rkey %p to %s cb %p",
-                  opcode, buffer,
+                  "datatype %"PRIx64" remote_addr %"PRIx64
```
0x%"PRIx64

hiroyuki-sato replied:
done

hiroyuki-sato commented on src/ucp/tag/tag_match.c:
```diff
@@ -142,7 +142,7 @@ ucp_tag_exp_search_all(ucp_tag_match_t *tm, ucp_request_queue_t *req_queue,
     ucs_assertv((hash_sn == ULONG_MAX) && (wild_sn == ULONG_MAX),
-                "hash_seq=%lu wild_seq=%lu", hash_sn, wild_sn);
+                "hash_seq=%"PRIu64" wild_seq=%"PRIu64, hash_sn, wild_sn);
     ucs_assert(ucs_queue_iter_end(hash_queue, hash_iter));
```
```C
    uint64_t hash_sn, wild_sn, *sn_p;
```

hiroyuki-sato commented on src/ucp/tag/tag_send.c:
```diff
@@ -46,7 +46,7 @@ ucp_tag_get_rndv_threshold(const ucp_request_t *req, size_t count,
     default:
-        ucs_error("Invalid data type %lx", req->send.datatype);
+        ucs_error("Invalid data type %"PRIx64, req->send.datatype);
     }
```
```C
ucp_tag_get_rndv_threshold(const ucp_request_t *req, size_t count,
                           size_t max_iov, size_t rndv_rma_thresh,
                           size_t rndv_am_thresh)

typedef struct ucp_request              ucp_request_t;


struct ucp_request {
        //....
        /* "receive" part - used for tag_recv and stream_recv operations */
        struct {
            //
            ucp_datatype_t        datatype; /* Receive type */
        } recv;

typedef uint64_t                         ucp_datatype_t;
```

yosefe commented on src/ucp/tag/tag_send.c:
```diff
@@ -46,7 +46,7 @@ ucp_tag_get_rndv_threshold(const ucp_request_t *req, size_t count,
         return rndv_am_thresh;
     default:
-        ucs_error("Invalid data type %lx", req->send.datatype);
+        ucs_error("Invalid data type %"PRIx64, req->send.datatype);
```
0x

hiroyuki-sato replied:
done

yosefe commented on src/ucp/tag/tag_send.c:
```diff
@@ -85,8 +85,8 @@ ucp_tag_send_req(ucp_request_t *req, size_t dt_count,
 
-    ucs_trace_req("select tag request(%p) progress algorithm datatype=%lx "
-                  "buffer=%p length=%zu max_short=%zd rndv_thresh=%zu "
+    ucs_trace_req("select tag request(%p) progress algorithm datatype=%"PRIx64
```
0x

hiroyuki-sato replied:
done

yosefe commented on src/ucp/wireup/address.c:
```diff
@@ -787,7 +787,9 @@ static ucs_status_t ucp_address_do_pack(ucp_worker_h worker, ucp_ep_h ep,
                ucs_trace("pack addr[%d] : "UCT_TL_RESOURCE_DESC_FMT" "
                           "eps %u md_flags 0x%"PRIx64" tl_flags 0x%"PRIx64" bw %e + %e/n ovh %e "
-                          "lat_ovh %e dev_priority %d a32 0x%lx/0x%lx a64 0x%lx/0x%lx",
+                          "lat_ovh %e dev_priority %d a32 "
+                          "0x%"PRIx64"/0x%"PRIx64" a64 "
+                          "0x%"PRIx64"/0x%"PRIx64,
```
wrap lines

hiroyuki-sato replied:
Could you check again? I wrapped by 80 characters.

yosefe commented on src/ucp/wireup/address.c:
```diff
@@ -1040,7 +1042,8 @@ ucs_status_t ucp_address_unpack(ucp_worker_t *worker, const void *buffer,
                           " tl_iface_flags 0x%"PRIx64" tl_event_flags 0x%"PRIx64
                           " bw %e + %e/n ovh %e lat_ovh %e dev_priority %d a32 "
-                          "0x%lx/0x%lx a64 0x%lx/0x%lx",
+                          "0x%"PRIx64"/0x%"PRIx64 
+                          " a64 0x%"PRIx64"/0x%"PRIx64,
```
wrap lines

hiroyuki-sato replied:
Could you check again? I wrapped by 80 characters.

hiroyuki-sato commented on src/ucp/wireup/address.c:
```diff
@@ -787,7 +787,9 @@ static ucs_status_t ucp_address_do_pack(ucp_worker_h worker, ucp_ep_h ep,
+                          "lat_ovh %e dev_priority %d a32 "
+                          "0x%"PRIx64"/0x%"PRIx64" a64 "
+                          "0x%"PRIx64"/0x%"PRIx64,
                           addr_index,
```
```C
    uct_iface_attr_t *iface_attr;

typedef struct uct_iface_attr      uct_iface_attr_t;



/**
 * @ingroup UCT_RESOURCE
 * @brief Interface attributes: capabilities and limitations.
 */
struct uct_iface_attr {
        //...
        struct {
            uint64_t         op_flags;   /**< Attributes for atomic-post operations */
            uint64_t         fop_flags;  /**< Attributes for atomic-fetch operations */
        } atomic32, atomic64;            /**< Attributes for atomic operations */
};
```

hiroyuki-sato commented on src/ucp/wireup/address.c:
```diff
@@ -1040,7 +1042,8 @@ ucs_status_t ucp_address_unpack(ucp_worker_t *worker, const void *buffer,
-                          "0x%lx/0x%lx a64 0x%lx/0x%lx",
+                          "0x%"PRIx64"/0x%"PRIx64 
+                          " a64 0x%"PRIx64"/0x%"PRIx64,
                           (int)(address - address_list), address->num_ep_addrs,
```
Same as above

hiroyuki-sato commented on src/ucp/wireup/select.c:
```diff
@@ -1503,7 +1503,7 @@ ucp_wireup_search_lanes(const ucp_wireup_select_params_t *select_params,
-        ucs_error("No transports selected to %s (features: 0x%lx)",
+        ucs_error("No transports selected to %s (features: 0x%"PRIx64")",
                   select_params->address->name,
                   ucp_ep_get_context_features(select_params->ep));
```
```C
static uint64_t ucp_ep_get_context_features(const ucp_ep_h ep)
{
    return ep->worker->context->config.features;
}
```

hiroyuki-sato commented on src/ucp/wireup/wireup.c:
```diff
@@ -924,7 +924,7 @@ static void ucp_wireup_print_config(ucp_worker_h worker,
     ucs_log(log_level,
-            "%s: am_lane %s wireup_lane %s cm_lane %s reachable_mds 0x%lx",
+            "%s: am_lane %s wireup_lane %s cm_lane %s reachable_mds 0x%"PRIx64,
             title, ucp_wireup_get_lane_index_str(key->am_lane, am_lane_str,
```
```C
    ucs_log(log_level,
            "%s: am_lane %s wireup_lane %s cm_lane %s reachable_mds 0x%"PRIx64,
            //...
            key->reachable_md_map);

static void ucp_wireup_print_config(/* ... */,
                                    const ucp_ep_config_key_t *key,
                                    /* ... */)

typedef struct ucp_ep_config_key        ucp_ep_config_key_t;

struct ucp_ep_config_key {
  //...
    /* Bitmap of remote mds which are reachable from this endpoint (with any set
     * of transports which could be selected in the future).
     */
    ucp_md_map_t             reachable_md_map;
}


UCP_UINT_TYPE(UCP_MD_INDEX_BITS)     ucp_md_map_t;

#define UCP_UINT_TYPE(_bits)         typedef UCS_PP_TOKENPASTE3(uint, _bits, _t)

```

hiroyuki-sato commented on src/ucp/wireup/wireup_cm.c:
```diff
@@ -219,8 +219,8 @@ static ssize_t ucp_cm_client_priv_pack_cb(void *arg,
-              dev_name, dev_index, tl_bitmap);
+    ucs_debug("client ep %p created on device %s idx %d, tl_bitmap 0x%"PRIx64,
+              ep, dev_name, dev_index, tl_bitmap);
     /* Pass real ep (not tmp_ep), because only its pointer and err_mode is
```
```C
    uint64_t tl_bitmap;
```

yosefe commented on src/ucp/proto/rndv.c:
```diff
@@ -763,7 +763,7 @@ UCS_PROFILE_FUNC_VOID(ucp_rndv_recv_frag_get_completion, (self, status),
 
     ucs_trace_req("freq:%p: recv_frag_get done. rreq:%p length:%ld"
-                  " offset:%"PRIu64,
+                  " offset:%"PRIx64,
```
PRIu64

yosefe commented:
```
  CXX      common/gtest-test_helpers.o
  CXX      common/gtest-test_obj_size.o
In file included from /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/rma/rma.h:11,
                 from /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_request.h:22,
                 from /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test_obj_size.cc:16:
/scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_rkey.h:97:23: error: invalid suffix on literal; C++11 requires a space between literal and string macro [-Werror=literal-suffix]
                       "(remote md_map 0x%"PRIx64")", \
                       ^
cc1plus: all warnings being treated as errors
make[3]: *** [common/gtest-test_obj_size.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: Leaving directory `/scrap/azure/agent-01/AZP_WORKSPACE/2/s/build-test/test/gtest'
```

yosefe commented:
```
[2020-08-10T13:09:50.431Z]                  from ../../../../test/gtest/ucp/test_ucp_dt.cc:7:
[2020-08-10T13:09:50.431Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucp/dt/dt.inl: In function ‘ucs_status_t ucp_dt_unpack_only(ucp_worker_h, void*, size_t, ucp_datatype_t, ucs_memory_type_t, const void*, size_t, int)’:
[2020-08-10T13:09:50.431Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucp/dt/dt.inl:110:45: error: expected ‘)’ before ‘PRIx64’
[2020-08-10T13:09:50.431Z]          ucs_fatal("unexpected datatype=0x%" PRIx64, datatype);
```

hiroyuki-sato commented:
@yosefe Thank you for your comment. I tried to fix that.

hiroyuki-sato commented:
Hello, @yosefe. Could you tell me MellanoxLab New failure results?
I'm not sure why CI failed in this environment. Do you know what OS version is used in MellanoxLab New?

hiroyuki-sato commented on src/ucp/core/ucp_request.c:
```diff
@@ -286,8 +288,8 @@ UCS_PROFILE_FUNC_VOID(ucp_request_memory_dereg, (context, datatype, state, req_d
-                   state);
+    ucs_trace_func("context=%p datatype=0x%"PRIu64 " state=%p", context,
+                   datatype, state);
 
```
Changed to `PRIx64`.

yosefe commented:
```
[2020-08-11T04:38:40.727Z]   CXX      ucp/gtest-test_ucp_peer_failure.o
[2020-08-11T04:38:44.285Z]   CXX      ucp/gtest-test_ucp_atomic.o
[2020-08-11T04:38:48.797Z]   CXX      ucp/gtest-test_ucp_dt.o
[2020-08-11T04:38:49.073Z] In file included from /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucs/sys/compiler.h:17:0,
[2020-08-11T04:38:49.073Z]                  from /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucs/arch/x86_64/cpu.h:11,
[2020-08-11T04:38:49.073Z]                  from /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucs/arch/cpu.h:90,
[2020-08-11T04:38:49.073Z]                  from /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucs/time/time.h:10,
[2020-08-11T04:38:49.073Z]                  from ../../../../test/gtest/common/test_helpers.h:20,
[2020-08-11T04:38:49.073Z]                  from ../../../../test/gtest/common/test.h:16,
[2020-08-11T04:38:49.073Z]                  from ../../../../test/gtest/ucp/test_ucp_dt.cc:7:
[2020-08-11T04:38:49.073Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucp/dt/dt.inl: In function ‘ucs_status_t ucp_dt_unpack_only(ucp_worker_h, void*, size_t, ucp_datatype_t, ucs_memory_type_t, const void*, size_t, int)’:
[2020-08-11T04:38:49.073Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucp/dt/dt.inl:111:45: error: expected ‘)’ before ‘PRIx64’
[2020-08-11T04:38:49.073Z]          ucs_fatal("unexpected datatype=0x%" PRIx64, datatype);
[2020-08-11T04:38:49.073Z]                                              ^
[2020-08-11T04:38:49.073Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucs/debug/assert.h:51:38: note: in definition of macro ‘ucs_fatal’
[2020-08-11T04:38:49.073Z]                             "Fatal: " _fmt, ## __VA_ARGS__)
[2020-08-11T04:38:49.073Z]                                       ^
[2020-08-11T04:38:49.073Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucs/debug/assert.h:51:58: error: spurious trailing ‘%’ in format [-Werror=format=]
[2020-08-11T04:38:49.073Z]                             "Fatal: " _fmt, ## __VA_ARGS__)
[2020-08-11T04:38:49.073Z]                                                           ^
[2020-08-11T04:38:49.073Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucp/dt/dt.inl:111:9: note: in expansion of macro ‘ucs_fatal’
[2020-08-11T04:38:49.073Z]          ucs_fatal("unexpected datatype=0x%" PRIx64, datatype);
[2020-08-11T04:38:49.073Z]          ^
[2020-08-11T04:38:49.073Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucs/debug/assert.h:51:58: error: too many arguments for format [-Werror=format-extra-args]
[2020-08-11T04:38:49.073Z]                             "Fatal: " _fmt, ## __VA_ARGS__)
[2020-08-11T04:38:49.073Z]                                                           ^
[2020-08-11T04:38:49.073Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucp/dt/dt.inl:111:9: note: in expansion of macro ‘ucs_fatal’
[2020-08-11T04:38:49.073Z]          ucs_fatal("unexpected datatype=0x%" PRIx64, datatype);
[2020-08-11T04:38:49.073Z]          ^
[2020-08-11T04:38:49.073Z] cc1plus: all warnings being treated as errors
[2020-08-11T04:38:49.073Z] make[4]: *** [ucp/gtest-test_ucp_dt.o] Error 1
[2020-08-11T04:38:49.073Z] make[4]: *** Waiting for unfinished jobs....
[2020-08-11T04:38:50.983Z] make[4]: Leaving directory `/scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/test/gtest'
[2020-08-11T04:38:50.983Z] make[3]: *** [all-recursive] Error 1
[2020-08-11T04:38:50.984Z] make[3]: Leaving directory `/scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/test/gtest'
[2020-08-11T04:38:50.984Z] make[2]: *** [all-recursive] Error 1
[2020-08-11T04:38:50.984Z] make[2]: Leaving directory `/scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub'
[2020-08-11T04:38:50.984Z] make[1]: *** [all] Error 2
[2020-08-11T04:38:50.984Z] make[1]: Leaving directory `/scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub'
[2020-08-11T04:38:50.984Z] make: *** [distcheck] Error 1
script returned exit code 2
```

yosefe commented:
```
08:45 [yosefe@r-vmb-ppc-jenkins ucx]$ uname -a
Linux r-vmb-ppc-jenkins 3.10.0-957.el7.ppc64le #1 SMP Thu Oct 4 20:51:36 UTC 2018 ppc64le ppc64le ppc64le GNU/Linux
08:45 [yosefe@r-vmb-ppc-jenkins ucx]$ cat /etc/redhat-release 
Red Hat Enterprise Linux Server release 7.6 (Maipo)
08:46 [yosefe@r-vmb-ppc-jenkins ucx]$ gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-36)
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

hiroyuki-sato commented:
@yosefe Thank you for your reply. I'll investigate it. But I successfully build this branch in my environment (CentOS7 x86_64).
and added `#include <inttypes.h>` explicitly in the same file. I'm not sure the root cause yet. (inline..?)

hoopoepg commented:
In C++ the macros are not automatically defined just by including the file.

You need to add the following:

```#define __STDC_FORMAT_MACROS 1```
before
```#include <inttypes.h>```

hiroyuki-sato commented:
@hoopoepg Thank you for your comment!
I added `__STDC_FORMAT_MACROS` macros in dt.c (Should define in anther place?)

Some platforms successfully build except PPC. It's strange for me
Anyway, thank you for your advice. It's very helpful.

yosefe commented:
still fails
let's add `#define __STDC_FORMAT_MACROS` and #include <inttypes.h> into test/gtest/common/test_.h
```
[2020-08-11T15:09:23.416Z]   CXX      ucp/gtest-test_ucp_memheap.o
[2020-08-11T15:09:23.705Z]   CXX      ucp/gtest-test_ucp_mmap.o
[2020-08-11T15:09:24.312Z] In file included from /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucs/debug/log.h:14:0,
[2020-08-11T15:09:24.312Z]                  from ../../../../test/gtest/common/test.h:18,
[2020-08-11T15:09:24.312Z]                  from ../../../../test/gtest/ucp/test_ucp_mmap.cc:7:
[2020-08-11T15:09:24.312Z] ../../../../test/gtest/ucp/test_ucp_mmap.cc: In member function ‘bool test_ucp_mmap::resolve_rma(ucp_test_base::entity*, ucp_rkey_h)’:
[2020-08-11T15:09:24.312Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucp/core/ucp_rkey.h:97:44: error: expected ‘)’ before ‘PRIx64’
[2020-08-11T15:09:24.312Z]                        "(remote md_map 0x%" PRIx64 ")", \
[2020-08-11T15:09:24.312Z]                                             ^
[2020-08-11T15:09:24.312Z] /scrap/jenkins/workspace/ucx/build-test/ucx-1.10.0/_build/sub/../../src/ucs/debug/log_def.h:35:75: note: in definition of macro ‘ucs_log_component’
[2020-08-11T15:09:24.312Z]                               (ucs_log_level_t)(_level), _comp_log_config, _fmt, ## __VA_ARGS__); \

```

yosefe commented on src/ucp/core/ucp_request.inl:
```diff
@@ -151,7 +151,7 @@ static UCS_F_ALWAYS_INLINE void
 {
     ucs_trace_req("completing receive request %p (%p) "UCP_REQUEST_FLAGS_FMT
-                  " stag 0x%"PRIx64" len %zu, %s",
+                  " stag 0x%" PRIx64 " len %zu, %s",
```
why needed to add spaces around PRIx64 ?

hiroyuki-sato replied:
For fixing this error.

https://github.com/openucx/ucx/pull/5552#issuecomment-671231079

```
  CXX      common/gtest-test_helpers.o
  CXX      common/gtest-test_obj_size.o
In file included from /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/rma/rma.h:11,
                 from /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_request.h:22,
                 from /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test_obj_size.cc:16:
/scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_rkey.h:97:23: error: invalid suffix on literal; C++11 requires a space between literal and string macro [-Werror=literal-suffix]
                       "(remote md_map 0x%"PRIx64")", \
                       ^
cc1plus: all warnings being treated as errors
make[3]: *** [common/gtest-test_obj_size.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: Leaving directory `/scrap/azure/agent-01/AZP_WORKSPACE/2/s/build-test/test/gtest'
```

yosefe commented on src/ucp/core/ucp_request.inl:
```diff
@@ -558,7 +558,7 @@ ucp_request_recv_data_unpack(ucp_request_t *req, const void *data,
 
     default:
-        ucs_fatal("unexpected datatype=%lx", req->recv.datatype);
+        ucs_fatal("unexpected datatype=0x%" PRIx64, req->recv.datatype);
```
remove space before PRIx64

hiroyuki-sato replied:
Same as the above reason. It seems that C++ requires a space between string literal and macro. IIUC this file is used by C++. correct?

yosefe commented on src/ucp/core/ucp_rkey.h:
```diff
@@ -93,7 +93,8 @@ typedef struct ucp_rkey {
         if (ucs_unlikely((_rkey)->cache._op_type##_lane == UCP_NULL_LANE)) { \
-            ucs_error("remote memory is unreachable (remote md_map 0x%lx)", \
+            ucs_error("remote memory is unreachable " \
+                      "(remote md_map 0x%" PRIx64 ")", \
```
remove space around PRIx64

yosefe commented on src/ucp/dt/dt.inl:
```diff
@@ -107,7 +108,7 @@ ucp_dt_unpack_only(ucp_worker_h worker, void *buffer, size_t count,
 
     default:
-        ucs_fatal("unexpected datatype=%lx", datatype);
+        ucs_fatal("unexpected datatype=0x%" PRIx64, datatype);
```
remove space around PRIx64

yosefe commented on src/ucp/proto/rndv.c:
```diff
@@ -1655,24 +1656,24 @@ static void ucp_rndv_dump(ucp_worker_h worker, uct_am_trace_type_t type,
         break;
     case UCP_AM_ID_RNDV_ATP:
-        snprintf(buffer, max, "RNDV_ATP sreq 0x%lx status '%s'",
+        snprintf(buffer, max, "RNDV_ATP sreq 0x%"PRIx64 " status '%s'",
```
remove space after PRIx64

yosefe commented on src/ucp/proto/rndv.c:
```diff
@@ -1655,24 +1656,24 @@ static void ucp_rndv_dump(ucp_worker_h worker, uct_am_trace_type_t type,
         break;
     case UCP_AM_ID_RNDV_DATA:
-        snprintf(buffer, max, "RNDV_DATA rreq 0x%"PRIx64" offset %zu",
+        snprintf(buffer, max, "RNDV_DATA rreq 0x%"PRIxPTR " offset %zu",
```
remove space after PRIxPTR

yosefe commented on src/ucp/proto/rndv.c:
```diff
@@ -1655,24 +1656,24 @@ static void ucp_rndv_dump(ucp_worker_h worker, uct_am_trace_type_t type,
-        snprintf(buffer, max, "RNDV_RTR sreq 0x%lx rreq 0x%lx address 0x%lx",
-                 rndv_rtr_hdr->sreq_ptr, rndv_rtr_hdr->rreq_ptr,
-                 rndv_rtr_hdr->address);
+        snprintf(buffer, max, "RNDV_RTR sreq 0x%"PRIxPTR " rreq 0x%"PRIxPTR
```
remove space after PRIxPTR

yosefe commented on src/ucp/proto/rndv.c:
```diff
@@ -1655,24 +1656,24 @@ static void ucp_rndv_dump(ucp_worker_h worker, uct_am_trace_type_t type,
         break;
     case UCP_AM_ID_RNDV_ATS:
-        snprintf(buffer, max, "RNDV_ATS sreq 0x%lx status '%s'",
+        snprintf(buffer, max, "RNDV_ATS sreq 0x%"PRIu64 " status '%s'",
```
remove space after PRIו64

hiroyuki-sato commented:
Hello, @yosefe. Thank you for your review.

I removed spaces and squashed.
The following files add space for avoiding C++ error.
These files are referenced by C++ files.

```C
src/ucp/core/ucp_rkey.h
97:                      "(remote md_map 0x%" PRIx64 ")", \

src/ucp/core/ucp_request.inl
154:                  " stag 0x%" PRIx64 " len %zu, %s",
561:        ucs_fatal("unexpected datatype=0x%" PRIx64, req->recv.datatype);

src/ucp/dt/dt.inl
111:        ucs_fatal("unexpected datatype=0x%" PRIx64, datatype);
```


As I said before, C++ requires space between literal and macros.

```C
#define __STDC_FORMAT_MACROS 1
#include <sys/types.h>
#include <stdio.h>
#include <inttypes.h>

int main(){
    uint64_t test = 12345;

    printf("%"PRIu64"\n",test);
    return(0);
}
```

```
c++ -Wall -std=c++11 test.cc
test.cc:9:15: error: invalid suffix on literal; C++11 requires a space between
      literal and identifier [-Wreserved-user-defined-literal]
    printf("%"PRIu64"\n",test);
              ^
```



It seems that space doesn't need after `PRIx64`.

* NG: `0x%"PRIx64")"`
* OK: `0x%" PRIx64 ")"`
* OK: `0x%" PRIx64")"`

Should I change like `0x%" PRIx64")"`?

yosefe commented:
> Should I change like `0x%" PRIx64")"`?

Yes, thank you!

hiroyuki-sato commented:
@yosefe Thank you for your review.
I removed spaces after `PRIx64`. Please take a look when you get a chance.

shamisp commented:
👍

hiroyuki-sato commented:
Thank you!
